packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,248 @@
+import { StaticDataView, sizeOfUTF8 } from './data-view';
+
+export type EnvKeys =
+  | 'ext_ghostery'
+  | 'ext_abp'
+  | 'ext_ublock'
+  | 'ext_ubol'
+  | 'ext_devbuild'
+  | 'env_chromium'
+  | 'env_edge'
+  | 'env_firefox'
+  | 'env_mobile'
+  | 'env_safari'
+  | 'env_mv3'
+  | 'false'
+  | 'cap_html_filtering'
+  | 'cap_user_stylesheet'
+  | 'adguard'
+  | 'adguard_app_android'
+  | 'adguard_app_ios'
+  | 'adguard_app_mac'
+  | 'adguard_app_windows'
+  | 'adguard_ext_android_cb'
+  | 'adguard_ext_chromium'
+  | 'adguard_ext_edge'
+  | 'adguard_ext_firefox'
+  | 'adguard_ext_opera'
+  | 'adguard_ext_safari'
+  | (string & {});
+
+export class Env extends Map<EnvKeys, boolean> {
+  constructor() {
+    super();
+
+    // Control the default conditions via constructor
+    this.set('true', true);
+    this.set('false', false);
+  }
+}
+
+export const enum PreprocessorTypes {
+  INVALID = 0,
+  BEGIF = 1,
+  ELSE = 2,
+  ENDIF = 3,
+}
+
+export function detectPreprocessor(line: string) {
+  // Minimum size of a valid condition should be 6 for something like: "!#if x" or "!#else"
+  if (
+    line.length < 6 ||
+    line.charCodeAt(0) !== 33 /* '!' */ ||
+    line.charCodeAt(1) !== 35 /* '#' */
+  ) {
+    return PreprocessorTypes.INVALID;
+  }
+
+  if (line.startsWith('!#if ')) {
+    return PreprocessorTypes.BEGIF;
+  }
+
+  if (line.startsWith('!#else')) {
+    return PreprocessorTypes.ELSE;
+  }
+
+  if (line.startsWith('!#endif')) {
+    return PreprocessorTypes.ENDIF;
+  }
+
+  return PreprocessorTypes.INVALID;
+}
+
+const tokenizerPattern = /(!|&&|\|\||\(|\)|[a-zA-Z0-9_]+)/g;
+const identifierPattern = /^!?[a-zA-Z0-9_]+$/;
+
+const tokenize = (expression: string) => expression.match(tokenizerPattern);
+const isIdentifier = (expression: string) => identifierPattern.test(expression);
+
+const precedence: Record<string, number> = {
+  '!': 2,
+  '&&': 1,
+  '||': 0,
+};
+
+const isOperator = (token: string) => Object.prototype.hasOwnProperty.call(precedence, token);
+
+const testIdentifier = (identifier: string, env: Env): boolean => !!env.get(identifier);
```

Comment:
Instead of using a custom class inheriting from `Map` above maybe it'd be simpler to do it here? This way we can use a normal `Map` which would make usage simpler.
```suggestion
const testIdentifier = (identifier: string, env: Env): boolean => {
  if (identifier === 'true') {
    return true;
  }
  
  if (identifier === 'false) {
    return false;
  }
  
  return !!env.get(identifier);
}
```

Reply:
This change was to address the use of ignoring all preprocessor conditions. If the user don't want any from the filters with preprocessors, just retrieving all values from env will allow the case.